File: tools/e2e_test.js
#!/usr/bin/env node
/**
 * LightSpeed E2E Tunnel Test (Node.js)
 *
 * Tests the full data path:
 *   Client (BKK) → Proxy (LA/OCI/SGP) → Game Server (SGP echo) → Proxy → Client
 *
 * Encodes LightSpeed tunnel headers matching protocol/src/header.rs format.
 */

const dgram = require('dgram');
const { Buffer } = require('buffer');
const { performance } = require('perf_hooks');

// ── Header format (20 bytes) ────────────────────────────────────
const PROTOCOL_VERSION = 1;
const HEADER_SIZE = 20;
const FLAG_KEEPALIVE = 0x01;

function ipToBytes(ip) {
  return Buffer.from(ip.split('.').map(Number));
}

function bytesToIp(buf, offset) {
  return `${buf[offset]}.${buf[offset+1]}.${buf[offset+2]}.${buf[offset+3]}`;
}

function nowUs() {
  return (Date.now() * 1000) >>> 0;
}

function encodeHeader(seq, timestampUs, srcIp, srcPort, dstIp, dstPort, flags = 0, sessionToken = 0) {
  const buf = Buffer.alloc(HEADER_SIZE);
  buf[0] = (PROTOCOL_VERSION << 4) | (flags & 0x0F);
  buf[1] = sessionToken;
  buf.writeUInt16BE(seq, 2);
  buf.writeUInt32BE(timestampUs, 4);
  ipToBytes(srcIp).copy(buf, 8);
  ipToBytes(dstIp).copy(buf, 12);
  buf.writeUInt16BE(srcPort, 16);
  buf.writeUInt16BE(dstPort, 18);
  return buf;
}

function decodeHeader(data) {
  if (data.length < HEADER_SIZE) throw new Error(`Too short: ${data.length} < ${HEADER_SIZE}`);
  const verFlags = data[0];
  return {
    header: {
      version: verFlags >> 4,
      flags: verFlags & 0x0F,
      sessionToken: data[1],
      sequence: data.readUInt16BE(2),
      timestampUs: data.readUInt32BE(4),
      srcIp: bytesToIp(data, 8),
      srcPort: data.readUInt16BE(16),
      dstIp: bytesToIp(data, 12),
      dstPort: data.readUInt16BE(18),
    },
    payload: data.slice(HEADER_SIZE),
  };
}

function testProxy(name, proxyHost, proxyPort, gameServerIp, gameServerPort) {
  return new Promise((resolve) => {
    console.log(`\n${'─'.repeat(50)}`);
    console.log(`Testing proxy: ${name} (${proxyHost}:${proxyPort})`);
    console.log(`Game server:   ${gameServerIp}:${gameServerPort}`);
    console.log('─'.repeat(50));

    const sock = dgram.createSocket('udp4');
    const results = { keepalive: null, rtts: [], errors: 0 };
    let phase = 'keepalive';
    let packetIdx = 0;
    const totalPackets = 5;
    let sentPayloads = [];
    let sendTimes = [];

    sock.on('error', (err) => {
      console.log(`  ❌ Socket error: ${err.message}`);
      results.errors++;
    });

    sock.on('message', (msg, rinfo) => {
      const recvTime = performance.now();

      try {
        const { header, payload } = decodeHeader(msg);

        if (phase === 'keepalive') {
          const rtt = recvTime - sendTimes[0];
          results.keepalive = rtt;
          console.log(`\n[1] Keepalive test:`);
          console.log(`  ✅ Keepalive echo: ${rtt.toFixed(1)}ms (seq=${header.sequence}, flags=0x${header.flags.toString(16).padStart(2,'0')})`);

          // Start data relay test
          phase = 'data';
          packetIdx = 0;
          sendTimes = [];
          sentPayloads = [];
          console.log(`\n[2] Data relay test (${totalPackets} packets):`);
          sendDataPacket();
          return;
        }

        if (phase === 'data') {
          const rtt = recvTime - sendTimes[packetIdx];
          results.rtts.push(rtt);

          const expectedPayload = sentPayloads[packetIdx];
          const match = payload.equals(expectedPayload);
          console.log(`  ${match ? '✅' : '❌'} Packet ${packetIdx}: ${rtt.toFixed(1)}ms, payload=${match ? 'MATCH' : 'MISMATCH'} (${payload.length}B), seq=${header.sequence}`);

          if (!match) {
            console.log(`      Sent:     ${expectedPayload.toString().slice(0, 40)}`);
            console.log(`      Received: ${payload.toString().slice(0, 40)}`);
          }

          packetIdx++;
          if (packetIdx < totalPackets) {
            setTimeout(sendDataPacket, 100);
          } else {
            finishTest();
          }
        }
      } catch (e) {
        console.log(`  ❌ Decode error: ${e.message}`);
        results.errors++;
      }
    });

    function sendDataPacket() {
      const payload = Buffer.from(`LIGHTSPEED_E2E_TEST_${packetIdx}_${Math.floor(Date.now()/1000)}`);
      sentPayloads[packetIdx] = payload;

      const header = encodeHeader(
        packetIdx, nowUs(),
        '127.0.0.1', 12345,
        gameServerIp, gameServerPort
      );

      const packet = Buffer.concat([header, payload]);
      sendTimes[packetIdx] = performance.now();
      sock.send(packet, proxyPort, proxyHost);

      // Timeout per packet
      setTimeout(() => {
        if (results.rtts.length <= packetIdx && phase === 'data') {
          console.log(`  ❌ Packet ${packetIdx}: TIMEOUT`);
          packetIdx++;
          if (packetIdx < totalPackets) {
            sendDataPacket();
          } else {
            finishTest();
          }
        }
      }, 5000);
    }

    let finished = false;
    function finishTest() {
      if (finished) return;
      finished = true;

      if (results.rtts.length > 0) {
        const avg = results.rtts.reduce((a,b) => a+b, 0) / results.rtts.length;
        const min = Math.min(...results.rtts);
        const max = Math.max(...results.rtts);
        console.log(`\n  📊 Results: avg=${avg.toFixed(1)}ms, min=${min.toFixed(1)}ms, max=${max.toFixed(1)}ms`);
        console.log(`  📊 Packets: ${results.rtts.length}/${totalPackets} received`);
        if (results.keepalive !== null) {
          const overhead = avg - results.keepalive;
          console.log(`  📊 Relay overhead vs keepalive: ${overhead >= 0 ? '+' : ''}${overhead.toFixed(1)}ms`);
        }
      } else {
        console.log(`\n  ❌ No responses received - relay may not be working`);
      }

      try { sock.close(); } catch(e) {}
      resolve(results);
    }

    // Start with keepalive test
    sock.bind(0, () => {
      const keepaliveHeader = encodeHeader(
        60000, nowUs(),
        '0.0.0.0', 0,
        '0.0.0.0', 0,
        FLAG_KEEPALIVE
      );

      sendTimes[0] = performance.now();
      sock.send(keepaliveHeader, proxyPort, proxyHost);

      // Keepalive timeout
      setTimeout(() => {
        if (results.keepalive === null) {
          console.log(`\n[1] Keepalive test:`);
          console.log(`  ❌ Keepalive: TIMEOUT (5s)`);

          // Still try data relay
          phase = 'data';
          packetIdx = 0;
          sendTimes = [];
          sentPayloads = [];
          console.log(`\n[2] Data relay test (${totalPackets} packets):`);
          sendDataPacket();
        }
      }, 5000);
    });
  });
}

async function main() {
  console.log('='.repeat(60));
  console.log('LightSpeed E2E Tunnel Test');
  console.log('='.repeat(60));

  const proxies = [
    ['Vultr LA',     '149.28.84.139',  4434],
    ['OCI San Jose', '163.192.3.134',  4434],
    ['Vultr SGP',    '149.28.144.74',  4434],
  ];
  const gameServerIp = '149.28.144.74';
  const gameServerPort = 9999;

  for (const [name, host, port] of proxies) {
    await testProxy(name, host, port, gameServerIp, gameServerPort);
  }

  console.log(`\n${'='.repeat(60)}`);
  console.log('Test complete');
  console.log('='.repeat(60));
}

main().catch(console.error);

File: tools/e2e_test.py
#!/usr/bin/env python3
"""
LightSpeed E2E Tunnel Test

Tests the full data path:
  Client (BKK) → Proxy (LA/SGP) → Game Server (SGP echo) → Proxy → Client

Encodes LightSpeed tunnel headers matching protocol/src/header.rs format.
"""
import socket
import struct
import time
import sys

# ── Header format (20 bytes) ────────────────────────────────────────
# Byte 0:  (version << 4) | (flags & 0x0F)
# Byte 1:  session_token
# Bytes 2-3:  sequence (big endian u16)
# Bytes 4-7:  timestamp_us (big endian u32)
# Bytes 8-11: orig_src_ip (4 bytes)
# Bytes 12-15: orig_dst_ip (4 bytes)
# Bytes 16-17: orig_src_port (big endian u16)
# Bytes 18-19: orig_dst_port (big endian u16)

PROTOCOL_VERSION = 1
HEADER_SIZE = 20
FLAG_KEEPALIVE = 0x01


def ip_to_bytes(ip_str):
    """Convert dotted IP string to 4 bytes."""
    return socket.inet_aton(ip_str)


def bytes_to_ip(b):
    """Convert 4 bytes to dotted IP string."""
    return socket.inet_ntoa(b)


def encode_header(seq, timestamp_us, src_ip, src_port, dst_ip, dst_port,
                  flags=0, session_token=0):
    """Encode a LightSpeed tunnel header."""
    ver_flags = (PROTOCOL_VERSION << 4) | (flags & 0x0F)
    return struct.pack('!BBHI',
                       ver_flags, session_token, seq, timestamp_us) + \
           ip_to_bytes(src_ip) + ip_to_bytes(dst_ip) + \
           struct.pack('!HH', src_port, dst_port)


def decode_header(data):
    """Decode a LightSpeed tunnel header, return (header_dict, payload)."""
    if len(data) < HEADER_SIZE:
        raise ValueError(f"Too short: {len(data)} < {HEADER_SIZE}")

    ver_flags, token, seq, ts = struct.unpack('!BBHI', data[:8])
    version = ver_flags >> 4
    flags = ver_flags & 0x0F

    src_ip = bytes_to_ip(data[8:12])
    dst_ip = bytes_to_ip(data[12:16])
    src_port, dst_port = struct.unpack('!HH', data[16:20])

    header = {
        'version': version,
        'flags': flags,
        'session_token': token,
        'sequence': seq,
        'timestamp_us': ts,
        'src_ip': src_ip,
        'src_port': src_port,
        'dst_ip': dst_ip,
        'dst_port': dst_port,
    }
    return header, data[HEADER_SIZE:]


def now_us():
    """Current time in microseconds (wrapping u32)."""
    return int(time.time() * 1_000_000) & 0xFFFFFFFF


def test_keepalive(proxy_addr):
    """Test keepalive echo."""
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.settimeout(5.0)

    header = encode_header(
        seq=60000,
        timestamp_us=now_us(),
        src_ip='0.0.0.0', src_port=0,
        dst_ip='0.0.0.0', dst_port=0,
        flags=FLAG_KEEPALIVE
    )

    t0 = time.perf_counter()
    sock.sendto(header, proxy_addr)

    try:
        data, addr = sock.recvfrom(2048)
        rtt_ms = (time.perf_counter() - t0) * 1000
        resp_hdr, _ = decode_header(data)
        print(f"  ✅ Keepalive echo: {rtt_ms:.1f}ms (seq={resp_hdr['sequence']}, "
              f"flags=0x{resp_hdr['flags']:02x})")
        return rtt_ms
    except socket.timeout:
        print(f"  ❌ Keepalive: TIMEOUT (5s)")
        return None
    finally:
        sock.close()


def test_data_relay(proxy_addr, game_server_ip, game_server_port, count=5):
    """Test full data relay through the proxy to a game server."""
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.settimeout(5.0)

    # Get our local IP (for the header)
    local_ip = '127.0.0.1'
    local_port = sock.getsockname()[1] or 12345

    rtts = []
    for i in range(count):
        payload = f"LIGHTSPEED_E2E_TEST_{i}_{time.time():.0f}".encode()
        seq = i

        header = encode_header(
            seq=seq,
            timestamp_us=now_us(),
            src_ip=local_ip, src_port=local_port,
            dst_ip=game_server_ip, dst_port=game_server_port,
        )

        packet = header + payload
        t0 = time.perf_counter()
        sock.sendto(packet, proxy_addr)

        try:
            data, addr = sock.recvfrom(2048)
            rtt_ms = (time.perf_counter() - t0) * 1000
            resp_hdr, resp_payload = decode_header(data)
            rtts.append(rtt_ms)

            payload_match = resp_payload == payload
            print(f"  {'✅' if payload_match else '❌'} Packet {i}: "
                  f"{rtt_ms:.1f}ms, "
                  f"payload={'MATCH' if payload_match else 'MISMATCH'} "
                  f"({len(resp_payload)}B), "
                  f"seq={resp_hdr['sequence']}")

            if not payload_match:
                print(f"      Sent:     {payload[:40]}")
                print(f"      Received: {resp_payload[:40]}")

        except socket.timeout:
            print(f"  ❌ Packet {i}: TIMEOUT")

        time.sleep(0.1)  # Don't flood

    sock.close()
    return rtts


def main():
    print("=" * 60)
    print("LightSpeed E2E Tunnel Test")
    print("=" * 60)

    # Configuration
    proxies = [
        ("Vultr LA",    ("149.28.84.139", 4434)),
        ("Vultr SGP",   ("149.28.144.74", 4434)),
    ]
    game_server_ip = "149.28.144.74"  # SGP echo server
    game_server_port = 9999

    for name, proxy_addr in proxies:
        print(f"\n{'─' * 50}")
        print(f"Testing proxy: {name} ({proxy_addr[0]}:{proxy_addr[1]})")
        print(f"Game server:   {game_server_ip}:{game_server_port}")
        print(f"{'─' * 50}")

        # Test 1: Keepalive
        print(f"\n[1] Keepalive test:")
        ka_rtt = test_keepalive(proxy_addr)

        # Test 2: Data relay
        print(f"\n[2] Data relay test (5 packets):")
        rtts = test_data_relay(proxy_addr, game_server_ip, game_server_port, count=5)

        if rtts:
            avg = sum(rtts) / len(rtts)
            mn = min(rtts)
            mx = max(rtts)
            print(f"\n  📊 Results: avg={avg:.1f}ms, min={mn:.1f}ms, max={mx:.1f}ms")
            print(f"  📊 Packets: {len(rtts)}/5 received")
            if ka_rtt:
                overhead = avg - ka_rtt
                print(f"  📊 Relay overhead vs keepalive: {overhead:+.1f}ms")
        else:
            print(f"\n  ❌ No responses received - relay may not be working")

    print(f"\n{'=' * 60}")
    print("Test complete")
    print(f"{'=' * 60}")


if __name__ == '__main__':
    main()
